### ISSUE TITLE
API benchopt run -f baseline reinstalls all solvers

### ISSUE THREAD

**Issue by mathurinm** [benchopt/benchOpt#16]
ad then it runs them all:

```
(base) ➜  benchOpt git:(add_fista) ✗ benchopt run lasso -f baseline -n 1 -d boston
Uninstalling Baseline in lasso:... done
Installing Baseline in lasso:... done
Installing Blitz in lasso:... done
Installing cd in lasso:... done
Installing Celer in lasso:... done
Installing cvxpy in lasso:... done
Installing Cyanure in lasso:... done
Installing Lightning in lasso:... done
Boston
|--Lasso regression(reg=0.05)
|----Baseline(use_acceleration=false): done               
|----Baseline(use_acceleration=false): done               
|----Blitz: done  
```

**Comment by tomMoral** [comment]
Any call with only `-f solver` will run all the solvers. This is because by default, we run all the solvers and adding `-f` just force to reinstall this solver and then run all.

Maybe we should change this to have `-f` only run one solver. But then if you want to run all solvers and reinstall one, this is a bit more complicated, not sure what is the right solution here

**Comment by agramfort** [comment]
I am fine to current behavior as long as it's well explained in the doc.

to reinstall and run you can do:

`benchopt run lasso -f baseline -n 1 -d boston -s baseline`

an alternative to do

`benchopt run lasso -s baseline -n 1 -d boston -f`

which could mean run only baseline and reinstall all tested solvers. it depends how likely it is to reinstall a solver and want more than this solver.

**Comment by mathurinm** [comment]
@agramfort  don't you think if I want to force the re-run of my solver, I should not have to wait 5 minutes for the installation of packages I don't want ?


Also @tomMoral I expected from the -s/-f parallel syntax that the only difference is to force rerun. If we keep it this way I'd add to the doc "all solvers are still benchmarked" in:
```
  -f, --force-solver <solver_name>
                                  Force the re-installation and run for
                                  <solver_name>. This avoids caching effect
                                  when adding an estimator.
```
 

IMO as more solvers will be added added the behavior will be to not run all solvers by default, but only the one specified.

**Comment by agramfort** [comment]
>
> @agramfort <https://github.com/agramfort> don't you think if I want to
> force the re-run of my solver, I should not have to wait 5 minutes for the
> installation of packages I don't want ?
>

true. Current API makes it hard?


> Also @tomMoral <https://github.com/tomMoral> I expected from the -s/-f
> parallel syntax that the only difference is to force rerun. If we keep it
> this way I'd add to the doc "all solvers are still benchmarked" in:
>
>   -f, --force-solver <solver_name>
>                                   Force the re-installation and run for
>                                   <solver_name>. This avoids caching effect
>                                   when adding an estimator.
>
> IMO as more solvers will be added the behavior will be to not run all
> solvers by default, but only the ones specified.
>
you can specify the ones you want with -s
we could also have -o for "omit" a specific one

**Comment by tomMoral** [comment]
I actually agree with @mathurinm  here that most people won't run all solvers, but only the competitive ones. My take on this:

* `-f` do the same as `-s` but force the reinstall
* `-s all` add all the solvers, so you can still do all easily once you specify extra options.
* `-x solver_name` remove one solver from the list (to all all except one).

Not a big fan of `-o, --omit` because it feels like output but I am open.
WDYT?

**Comment by agramfort** [comment]
> -f do the same as -s but force the reinstall

ok

> -s all add all the solvers, so you can still do all easily once you specify extra options.

isn't "-s all" like now not passing any -s ?

> -x solver_name remove one solver from the list (to all all except one).

ok

> Not a big fan of -o, --omit because it feels like output but I am open.
> WDYT?

+1 for +x for "exclude"

**Comment by tomMoral** [comment]
> isn't "-s all" like now not passing any -s ?

Yes but it allows you to pass `-f` for only one solver while still running all th other and also allow including all and then using `-x` to exclude one.

**Comment by agramfort** [comment]
writing the doc for this will tell us how convoluted this is.

maybe -a for all? i don't like much that a solver name "all" is a special
case.

>

**Comment by tomMoral** [comment]
The only issue I see with `-a` is that it could apply for solvers or datasets. But the same apply with `-f` or `-x`.
You are probably right that `-s all` is not perfect.

**Comment by agramfort** [comment]
maybe "-s *" if it's doable or -s without value means all?

**Comment by tomMoral** [comment]
I like `-s *` very much as `-s` is supposed to take regexp as input.

**Comment by agramfort** [comment]
deal !

**Comment by tomMoral** [comment]
@mathurinm do you want to play with implementing this? The dataset option `-d` alread implement the regex part so it would mean adapting the way we filter the solver to actually use regex matching and change the way we manage `exclude_solvers` and `force_solvers`.

It might not be the funniest thing so I can also do it and you could review and help me write the doc so we are on the same page for the API.

**Comment by mathurinm** [comment]
Yes I'd like to do it

**Comment by mathurinm** [comment]
In fact I won't have time to do it in the next 7 and it's a major pain when running on large datasets so if you can do it before that is also ok

**Comment by tomMoral** [comment]
Closing as this has been implemented in #55 .
